core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,575 @@
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Transactional overlay implementation.
+//!
+//! This follows a linear succession of states.
+//! This contains multiple unbounded transaction layer
+//! and an additional top level 'prospective' layer.
+//! It only allows linear history (no branch so
+//! inner storage is only an array of element).
+
+use rstd::vec::Vec;
+use rstd::vec;
+
+#[derive(Debug, Clone, Eq, PartialEq, Copy)]
+/// State of a transactional layer.
+pub enum TransactionState {
+	/// Data is under change and can still be dropped.
+	Pending,
+	/// Same as pending but does count as a transaction start.
+	TxPending,
+	/// Data pointing to this indexed historic state should
+	/// not be returned and can be removed.
+	Dropped,
+}
+
+/// An entry at a given history height.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test-helpers"), derive(PartialEq))]
+pub struct HistoricalValue<V> {
+	/// The stored value.
+	pub value: V,
+	/// The moment in history when the value got set.
+	pub index: usize,
+}
+
+impl<V> From<(V, usize)> for HistoricalValue<V> {
+	fn from(input: (V, usize)) -> HistoricalValue<V> {
+		HistoricalValue { value: input.0, index: input.1 }
+	}
+}
+
+impl<V> HistoricalValue<V> {
+	fn as_ref(&self) -> HistoricalValue<&V> {
+		HistoricalValue {
+			value: &self.value,
+			index: self.index,
+		}
+	}
+}
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+type MemoryOnly<V> = smallvec::SmallVec<[HistoricalValue<V>; ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+const ALLOCATED_HISTORY: usize = 2;
+
+/// History of value that are related to a state history (eg field `history` of
+/// an `OverlayedChangeSet`).
+///
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test-helpers"), derive(PartialEq))]
+pub struct History<V>(MemoryOnly<V>);
+
+impl<V> Default for History<V> {
+	fn default() -> Self {
+		History(Default::default())
+	}
+}
+
+// Following implementation are here to isolate
+// buffer specific functions.
+impl<V> History<V> {
+
+	fn get_state(&self, index: usize) -> HistoricalValue<&V> {
```

Comment:
Yes, this was intentional (it replaces some [] access to the inner vec), but code can be change to use 'get'.
Generally private function here are mostly running over expectations of correct usage, I will try to add some panic comment in the doc, since there is may.